test/jdk/java/text/Format/MessageFormat/MessageFormatToPatternTest.java

```diff
@@ -0,0 +1,323 @@
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @summary Verify MessageFormat.toPattern() is equivalent to original pattern
+ * @bug 8323699
+ * @run junit MessageFormatToPatternTest
+ */
+
+import java.text.ChoiceFormat;
+import java.text.DateFormat;
+import java.text.DecimalFormat;
+import java.text.Format;
+import java.text.MessageFormat;
+import java.text.NumberFormat;
+import java.text.SimpleDateFormat;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.Locale;
+import java.util.Random;
+import java.util.stream.Stream;
+
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class MessageFormatToPatternTest {
+
+    private static final int NUM_RANDOM_TEST_CASES = 1000;
+    private static final int MAX_FORMAT_NESTING = 3;
+
+    private static Locale savedLocale;
+    private static long randomSeed;             // set this to a non-zero value for reproducibility
+    private static Random random;
+    private static boolean spitSeed;
+    private static int textCount;
+
+// Setup & Teardown
+
+    @BeforeAll
+    public static void setup() {
+        savedLocale = Locale.getDefault();
+        Locale.setDefault(Locale.US);
+        if (randomSeed == 0)
+            randomSeed = new Random().nextLong();
+        random = new Random(randomSeed);
+    }
+
+    @AfterAll
+    public static void teardown() {
+        Locale.setDefault(savedLocale);
+    }
+
+// Tests
+
+    @Test
+    public void testJDK_8323699() {
+
+        // This is the test case from JDK-8323699
+        testRoundTrip(new MessageFormat("{2,choice,0.0#option A: {0}|1.0#option B: {0}'}'}"));
+
+        // A few more test cases from the PR#17416 discussion
+        testRoundTrip(new MessageFormat("Test: {0,number,foo'{'#.00}"));
+        testRoundTrip(new MessageFormat("Test: {0,number,foo'}'#.00}"));
+    }
+
+    // Go roundrip from MessageFormat -> pattern string -> MessageFormat and verify equivalence
+    @ParameterizedTest
+    @MethodSource("testCases")
+    public void testRoundTrip(MessageFormat format1) {
```

Comment:
Can we also include the original String pattern that created format1, to help with debugging. I find myself wondering what the original String was.

Since technically, the full round trip is _pattern string -> MessageFormat1 -> pattern string -> MessageFormat2_

Reply:
I would have done that but it's not (easily) possible. The `MessageFormat`'s are created not from format strings, but by piecing together plain text and sub-`Format` objects manually. This was we are sure what we're dealing with.

Trying to create format strings with multiple levels of nesting from scratch is too complex for my brain due to all the levels of quoting required.

Reply:
Right, should have noted that, definitely not worth to try and re-synthesize the original String pattern for each Format. If we're adding some additional concrete cases, its fine since those will clearly have the original String patterns there.